optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/solution/descriptor/ClassBrowser.java

```diff
@@ -0,0 +1,55 @@
+package org.optaplanner.core.impl.domain.solution.descriptor;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+import java.util.*;
+
+public class ClassBrowser {
+
+    private static Set<Class<?>> getAllClassesInHierarchy(Class<?> root) {
+        Set<Class<?>> classes = new HashSet<>();
+        classes.add(root);
+        Class<?> superClass = root.getSuperclass();
+        while (superClass != Object.class) {
+            classes.add(superClass);
+            superClass = superClass.getSuperclass();
+        }
+        return classes;
+    }
+
+    /**
+     *
+     * @param root Never null, the class in which to start the search.
+     * @return Never null; all fields in the class, and all non-private non-static fields in super-types.
+     */
+    public static Set<Field> getAllVisibleFields(Class<?> root) {
+        Set<Class<?>> classes = getAllClassesInHierarchy(root);
+        Set<Field> visibleFields = new HashSet<>();
+        classes.forEach(c -> Arrays.stream(c.getDeclaredFields()).forEach(f -> {
+            int mods = f.getModifiers();
+            if (!Modifier.isStatic(mods) && (c == root || !Modifier.isPrivate(mods))) {
```

Comment:
@ge0ffrey You were mentioning that SolutionDescriptor should fail fast if it finds an annotated fact/entity getter with non-zero number of attributes. And you noticed that AbstractSolution does just that, yet Planner doesn't fail. I find out it is because of this code.

This code makes sure that, in parent classes of a class (solution, in this case), it never returns a field or method which is not visible. (```private``` field/method from a super-type will not be visible.) This was exactly the case with ```AbstractSolution``` - the code always worked, and it worked as it was designed.

Just to be sure, I've created and pushed ```SolutionDescriptorTest``` verifying that this situation works out as it should.